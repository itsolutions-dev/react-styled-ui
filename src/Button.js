import styled from 'styled-components';
import {
  getBackgroundColor,
  getTextColor,
  getOpacity,
  getBorder,
  getHover,
} from './utils/';

const Button = styled.button`
  padding: 6px 12px;
  margin-left:5px;
  font-size: ${props => props.theme.font.size};
  background-color: ${props =>
    props.reverse === true ? getTextColor(props) : getBackgroundColor(props)};
  color: ${props =>
    props.reverse === true ? getBackgroundColor(props) : getTextColor(props)};
  opacity: ${props => getOpacity(props)};
  border: ${props =>
    props.reverse === true ? '0' : `2px solid ${getBorder(props)}`};
  &:hover {
    ${props => props.reverse === true ? `color: ${getHover(props)}` : ''};
    ${props => props.reverse === true ? '' : `background-color: ${getHover(props)}`};
  }
  &:focus {
    outline:0;
  }
`;

export default Button;
