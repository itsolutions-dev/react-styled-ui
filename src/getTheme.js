import defaultTheme from './theme';
import deepMerge from './utils/deepMerge';

const getTheme = deepMerge.bind(null, defaultTheme);

export default getTheme;
