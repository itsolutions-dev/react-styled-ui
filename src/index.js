export Absolute from './Absolute';
export Avatar from './Avatar';
export BackgroundImage from './BackgroundImage';
export Badge from './Badge';
export Blockquote from './Blockquote';
export Border from './Border';
export Button from './Button';
export ButtonCircle from './ButtonCircle';
export ButtonGroup from './ButtonGroup';
export Card from './Card';
export Checkbox from './Checkbox';
export Code from './Code';
export Container from './Container';
export Divider from './Divider';
export Fixed from './Fixed';
export Footer from './Footer';
export getTheme from './getTheme';
export Heading from './Heading';
export Heading1 from './Heading1';
export Heading2 from './Heading2';
export Heading3 from './Heading3';
export Heading4 from './Heading4';
export Heading5 from './Heading5';
export Heading6 from './Heading6';
export Image from './Image';
export InlineText from './InlineText';
export Input from './Input';
export Label from './Label';
export Link from './Link';
export List from './List';
export ListItem from './ListItem';
export Loader from './Loader';
export Message from './Message';
export Overlay from './Overlay';
export Progress from './Progress';
export Radio from './Radio';
export Relative from './Relative';
export Select from './Select';
export Slider from './Slider';
export Text from './Text';
export Textarea from './Textarea';
export Toggle from './Toggle';
export Toolbar from './Toolbar';
export Tooltip from './Tooltip';
export Truncate from './Truncate';
export ThemeProvider from './ThemeProvider';
export theme from './theme';
