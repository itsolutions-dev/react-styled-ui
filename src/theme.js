export default {
  radius: 8,
  spacing: {
    margin: 8,
    padding: 8,
    iconSize: 24,
    borderSize: 1,
    shadowSize: 1,
  },
  font: {
    size: 16,
    weight: 200,
    lineHeight: 18,
    family: 'Roboto, sans-serif',
  },
  h1: {
    size: 32,
    weight: 900,
    lineHeight: 64,
  },
  h2: {
    size: 24,
    weight: 700,
    lineHeight: 32,
  },
  h3: {
    size: 18,
    weight: 500,
    lineHeight: 24,
  },
  selected: {
    size: 0,
    weight: 10,
    lineHeight: 0,
  },
  hover: {
    size: 0,
    weight: 10,
    lineHeight: 0,
  },
  transitions: {
    in: undefined,
    out: undefined,
  },
  palette: {
    default: 'green',
    defaultText: 'white',
    defaultDisabled: undefined,
    defaultHover: undefined,
    defaultSelected: undefined,
    primary: 'red',
    primaryText: 'white',
    primaryDisabled: undefined,
    primaryHover: undefined,
    primarySelected: undefined,
    secondary: 'blue',
    secondaryText: 'white',
    secondaryDisabled: undefined,
    secondaryHover: undefined,
    secondarySelected: undefined,
    disabled: 'lightgray',
    shadow: 'black',
  },
};
