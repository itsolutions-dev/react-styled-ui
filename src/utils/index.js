export deepMerge from './deepMerge';
export getOpacity from './getOpacity';
export getProperty from './getProperty';
export getHover from './getHover';
export getBackgroundColor from './getBackgroundColor';
export getTextColor from './getTextColor';
export getBorder from './getBorder';
export getHeight from './getHeight';
export hexToRgb from './hexToRgb';
export getColorWithAlpha from './getColorWithAlpha';
export getSizeOrDefault from './getSizeOrDefault';
export guid from './guid';
