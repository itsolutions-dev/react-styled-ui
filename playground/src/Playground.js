// @flow

import React from 'react';
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live';
import ThemeProvider from '../../src/ThemeProvider';
import theme from '../../src/theme';

const getCode = code => `
  <ThemeProvider theme={theme}>
    ${code}
  </ThemeProvider>
`.trim().replace(/^/, '\t');

type PlaygroundProps = {
  code: string,
  scope: Object
};

export default ({ code, scope, ...others }: PlaygroundProps) =>
  (<LiveProvider
    code={getCode(code)}
    scope={{
      ThemeProvider,
      theme,
      ...scope,
    }}
    {...others}
  >
    <LiveEditor />
    <LiveError />
    <LivePreview />
  </LiveProvider>);
