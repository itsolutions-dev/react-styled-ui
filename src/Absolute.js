import styled, { css } from './styledComponent';
import { getSizeOrDefault } from './utils/';

const Absolute = styled({
  component: 'div',
  style: css`
    position: absolute;
    ${props => props.top !== undefined ? `top: ${getSizeOrDefault(props.top, 0)};` : ''}
    ${props => props.bottom !== undefined ? `bottom: ${getSizeOrDefault(props.bottom, 0)};` : ''}
    ${props => props.left !== undefined ? `left: ${getSizeOrDefault(props.left, 0)};` : ''}
    ${props => props.right !== undefined ? `right: ${getSizeOrDefault(props.right, 0)};` : ''}
  `,
});

export default Absolute;
