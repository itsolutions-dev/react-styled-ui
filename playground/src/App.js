import React from 'react';
import styled from 'styled-components';
import Playground from './Playground';
import * as components from '../../src/';
import ThemeProvider from '../../src/ThemeProvider';

const { Heading } = components;

const Body = styled.div`
  background: #FFFFFF;
`;

const filePathToComponentName = file =>
  file.replace(/\.\/(examples\/)?/, '').replace(/\.example$/, '');

const getExamples = () => {
  const req = require.context('./examples', true, /\.example$/);
  return req.keys().map((file) => {
    const code = req(file);
    const scope = [];

    let match;
    const componentNameRegex = /<([A-Z]\w*)[\s|>]/g;
    // eslint-disable-next-line
    while ((match = componentNameRegex.exec(code)) !== null) {
      scope.push({
        name: match[1],
        component: components[match[1]],
      });
    }

    return {
      name: filePathToComponentName(file),
      scope: scope.filter(x => x.component)
        .reduce((acc, cur) => ({
          ...acc,
          [cur.name]: cur.component,
        }), {}),
      code,
    };
  });
};

export default () => (
  <ThemeProvider>
    <Body>
      <Heading size={1}>Playground</Heading>
      {
        getExamples()
          // eslint-disable-next-line
          .sort((x, y) => x.name > y.name ? 1 : -1)
          .map(x => (
            <Playground
              key={x.name}
              code={x.code}
              scope={x.scope}
              collapsableCode
            />
          ))
      }
    </Body>
  </ThemeProvider>
);
