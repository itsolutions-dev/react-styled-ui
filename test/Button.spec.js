import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import 'styled-components-test-utils/lib/jest';
import theme from '../src/theme';
import getTheme from '../src/getTheme';
import Button from '../src/Button';
import * as utils from '../src/utils/';

describe('Button', () => {
  test('should render a Button', () => {
    const component = ReactTestRenderer.create(<Button theme={theme} />);
    expect(component).toBeDefined();
  });

  test('should have a button tag', () => {
    const component = ReactTestRenderer.create(<Button theme={theme} />).toJSON();
    expect(component.type).toEqual('button');
  });

  test('should have a background-color', () => {
    const spy = jest.spyOn(utils, 'getBackgroundColor');
    const component = ReactTestRenderer.create(
      <Button
        theme={theme}
        backgroundColor="white"
      />,
    );
    expect(component).toHaveStyleRule('background-color', 'white');
    expect(spy).toHaveBeenCalled();
  });

  test('should have a border', () => {
    const spy = jest.spyOn(utils, 'getBorder');
    const component = ReactTestRenderer.create(
      <Button
        theme={theme}
        borderColor="white"
      />,
    );
    expect(component).toHaveStyleRule('border', '2px solid white');
    expect(spy).toHaveBeenCalled();
  });

  test('should have a color', () => {
    const spy = jest.spyOn(utils, 'getTextColor');
    const component = ReactTestRenderer.create(
      <Button
        theme={theme}
        textColor="white"
      />,
    );
    expect(component).toHaveStyleRule('color', 'white');
    expect(spy).toHaveBeenCalled();
  });

  test('should have opacity', () => {
    const spy = jest.spyOn(utils, 'getOpacity');
    const component = ReactTestRenderer.create(
      <Button
        theme={theme}
        opacity={0.75}
      />,
    );
    expect(component).toHaveStyleRule('opacity', '0.75');
    expect(spy).toHaveBeenCalled();
  });

  test('should have a background-color hover', () => {
    const spy = jest.spyOn(utils, 'getHover');
    const component = ReactTestRenderer.create(
      <Button
        theme={theme}
        hoverColor="white"
      />,
    );
    expect({
      component,
      modifier: '&:hover',
    }).toHaveStyleRule('background-color', 'white');
    expect(spy).toHaveBeenCalled();
  });

  test('should be reversed', () => {
    const spies = [
      'getHover',
      'getTextColor',
      'getBackgroundColor',
      'getOpacity',
    ].map(x => jest.spyOn(utils, x));
    const component = ReactTestRenderer.create(
      <Button
        theme={getTheme({
          font: {
            size: 12,
          },
        })}
        hoverColor="red"
        textColor="green"
        backgroundColor="blue"
        reverse
      />,
    );
    expect(component).toHaveStyleRule('font-size', '12');
    expect(component).toHaveStyleRule('border', '0');
    expect(component).toHaveStyleRule('color', 'blue');
    expect(component).toHaveStyleRule('background-color', 'green');
    expect({
      component,
      modifier: '&:hover',
    }).toHaveStyleRule('color', 'red');
    spies.forEach((spy) => {
      expect(spy).toHaveBeenCalled();
    });
  });

  test('should have a radius', () => {
    const component = ReactTestRenderer.create(
      <Button
        theme={theme}
        radius={20}
      />,
    );
    expect(component).toHaveStyleRule('border-radius', '20px');
  });
});
