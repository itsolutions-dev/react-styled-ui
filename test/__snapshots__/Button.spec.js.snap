// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render a Button 1`] = `
.c0 {
  padding: 8px 16px;
  margin-left: 5px;
  margin-right: 5px;
  font-size: 16px;
  background: #ffffff;
  color: #4c4c4c;
  opacity: 1;
  border: 2px solid #bababa;
}

.c0:hover {
  background: #f5f7f7;
}

.c0:focus {
  outline: none;
}

<button
  className="c0"
/>
`;
